plugins/out_pgsql/pgsql.c

```diff
@@ -0,0 +1,268 @@
+        PQfreemem(temp);
+    }
+
+    flb_info("[out_pgsql] we check that the table %s exists, if not we create it", ctx->db_table);
+
+    str_len = 62 + flb_sds_len(ctx->db_table);
+
+    query = flb_malloc(str_len);
+    if(query == NULL) {
+        pgsql_conf_destroy(ctx);
+        return -1;
+    }
+
+    snprintf(query, str_len, "CREATE TABLE IF NOT EXISTS %s (data json);", ctx->db_table);
+    res = PQexec(ctx->conn, query);
+    flb_free(query);
+
+    if(PQresultStatus(res) != PGRES_COMMAND_OK) {
+        flb_error("[out_pgsql] %s", PQerrorMessage(ctx->conn));
+        pgsql_conf_destroy(ctx);
+        return -1;
+    }
+
+    PQclear(res);
+
+    flb_info("[out_pgsql] switching postgresql connection to non-blocking mode");
+    if(PQsetnonblocking(ctx->conn, 1) != 0) {
+        flb_error("[out_pgsql] non-blocking mode not set, not good");
+        pgsql_conf_destroy(ctx);
+        return -1;
+    }
+
+    return 0;
+}
+
+static void cb_pgsql_flush(const void *data, size_t bytes,
+                            const char *tag, int tag_len,
+                            struct flb_input_instance *i_ins,
+                            void *out_context,
+                            struct flb_config *config)
+{
+    struct flb_pgsql_config *ctx = out_context;
+    flb_sds_t json;
+    PGresult *res = NULL;
+    char *query = NULL;
+    size_t str_len;
+
+    if(PQconsumeInput(ctx->conn) == 0 && PQisBusy(ctx->conn) == 1) {
+        flb_debug("[out_pgsql] Some command may still be running");
+    }
+
+    /*
+      PQreset()
+      This function will close the connection to the server and attempt to reestablish a new connection
+      to the same server, using all the same parameters previously used. This might be useful for error
+      recovery if a working connection is lost.
+     */
+    if(PQstatus(ctx->conn) != CONNECTION_OK) {
+        PQreset(ctx->conn);
+        FLB_OUTPUT_RETURN(FLB_RETRY);
+    }
+
+
+    json = flb_pack_msgpack_to_json_format(data, bytes,
+                                           FLB_PACK_JSON_FORMAT_JSON,
+                                           FLB_PACK_JSON_DATE_DOUBLE,
+                                           ctx->timestamp_key);
+    if(json == NULL) {
+        FLB_OUTPUT_RETURN(FLB_RETRY);
+    }
+
+    str_len = 115 + flb_sds_len(json);
+    query = flb_malloc(str_len);
+
+    if (query == NULL) {
+        flb_sds_destroy(json);
+        FLB_OUTPUT_RETURN(FLB_RETRY);
+    }
+
+    snprintf(query, str_len,
+             "with json_pack as (select json_array_elements('%s') as json_pack )INSERT INTO %s select * from json_pack;",
+             json, ctx->db_table);
+
+    PQsendQuery(ctx->conn, query);
+    flb_free(query);
+    flb_sds_destroy(json);
+
+    PQflush(ctx->conn);
+
+    if(PQisBusy(ctx->conn) == 0) {
+        res = PQgetResult(ctx->conn);
+        if(PQresultStatus(res) != PGRES_COMMAND_OK) {
+            flb_warn("[out_pgsql] %s", PQerrorMessage(ctx->conn));
```

Comment:
Not required, but Eduardo recently added a new plugin logging API: https://github.com/fluent/fluent-bit/commit/96ecbcbed887b41ac59e861c2122b733bbd796cc

Reply:
Maybe we can leave it after the releases? we already changed a lot of stuff here, I can schedule a day to use that API, I think that would be great